mathics/core/atoms.py

```diff
@@ -387,29 +326,71 @@ class MachineReal(Real):
     Stored internally as a python float.
     """
 
+    # Collection of MachineReals defined so far.
```

Comment:
Suppose you evaluate something like this:
```
Min[Eigenvalues[Table[Random[], {10^6},{10^6}]]]
```
Then, after that, we are storing 10^12 useless random numbers, plus the corresponding hashes. 
(Usually in my work I have to compute things like this, but with a more complicated function instead of Random[], but the idea is the same....

Reply:
There may be some confusion here as well.  The only things that are stored in that dictionary are Python objects that were created anyway. And except in somewhat pathological situations, Python doesn't garbage collect these objects.

So in sum this is a problem we have independent of what we have here. If you run that code twice in the current system if it allocates 10^12 objects, it will allocate 10^12 _more_ the next time it is run without garbage collectiong the first 10^12. 

So what is needed independent of this is a better way to represent list literals and ensuring not to create Mathics objects like Integer except on demand.

Reply:
So, do you say that in master, the numbers in the array are not collected after the evaluation?

Reply:
Yes, I believe numbers are not generally garbage collected by Python. Prove me wrong though!

I believe that with this branch our memory consumption may go down. 

However, this situation even with this PR is far from ideal.  I know that there are some situations where we don't have to allocate integers. I recall making a change to the `Range()` class and I recall that was a noticeable win. 

And I think we do something like this in Plotting. But in general we need to regularize this kind of thing and do more of it.

Reply:
Good, that is something that I can check. In any case, to remove the singleton pattern in numbers should not be too hard.

Reply:
Here is the check. I hacked this branch to see what are the references for each element in the MachineReal instances dict:

https://github.com/Mathics3/mathics-core/tree/debug_MachineReal_references

Then I run this:

```
mathics -e 'Sqrt[124/743.]' -e 'Print["---"]'   -e '1.'
```

This is the last (and relevant) part of the output:

```

list of entries and their references for the dict (<class 'dict'>, 140479948858880)
0.0:              ['MachineReals._machine_reals', "dict: {'expr', 'atom'}", "dict: {'replace', 'system', 'pattern'}", "dict: {'expr', 'atom'}", "dict: {'expr', 'atom'}", "dict: {'expr', 'atom'}", "dict: {'expr', 'atom'}", "dict: {'expr', 'atom'}", "dict: {'expr', 'atom'}", 'class / module', 'class / module', 'class / module: \nRepresentation of Numbers\n\nIntegers and Real numb', 'tuple: (<MachineReal: 0.0>,)', 'tuple: (<MachineReal: 0.0>,)', 'tuple: (<MachineReal: 0.0>,)', 'tuple: (<MachineReal: 0.0>,)', 'tuple: (<Symbol: System`z>, <MachineReal: 0.0>)', 'tuple: (<MachineReal: 0.0>,)', 'tuple: (<MachineReal: 0.0>,)']
-3.0:             ['MachineReals._machine_reals', "dict: {'original', 'hash', '_value'}", "dict: {'replace', 'system', 'pattern'}", "class / module: \n    <dl>\n      <dt>'$SystemTimeZone'\n      <dd> g", "class / module: \n    <dl>\n      <dt>'$TimeZone'\n      <dd> gives t"]
0.6:              ['MachineReals._machine_reals', 'class / module: \nEvaluation routines for 2D plotting.\n\nThese routi', 'tuple: (<MachineReal: 0.6>, <MachineReal: 0.4>, <MachineReal: 0.2>)']
0.2:              ['MachineReals._machine_reals', 'class / module: \nEvaluation routines for 2D plotting.\n\nThese routi', 'tuple: (<MachineReal: 0.6>, <MachineReal: 0.4>, <MachineReal: 0.2>)', 'tuple: (<Symbol: System`i>, <Integer: 2>, <MachineReal: 0.2>)', 'tuple: (<Symbol: System`i>, <Symbol: System`r>, <MachineReal: 0.2>)']
-10.0:            ['MachineReals._machine_reals']
-1.0:             ['MachineReals._machine_reals']
0.5:              ['MachineReals._machine_reals', 'tuple: (<MachineReal: 0.5>,)', 'tuple: (<MachineReal: 0.5>,)', 'tuple: (<Integer: 1>, <MachineReal: 0.5>, <Integer: 0>)', 'tuple: (<MachineReal: 1.0>, <MachineReal: 0.5>, <MachineReal: 0.5>)', 'tuple: (<MachineReal: 0.5>, <Integer: 0>, <MachineReal: 0.5>)', 'tuple: (<Symbol: System`n>, <MachineReal: 0.5>)', 'tuple: (<Symbol: System`n>, <MachineReal: 0.5>)', 'tuple: (<Symbol: System`n>, <MachineReal: 0.5>)', 'tuple: (<Symbol: System`n>, <MachineReal: 0.5>)']
0.4:              ['MachineReals._machine_reals', 'tuple: (<MachineReal: 0.6>, <MachineReal: 0.4>, <MachineReal: 0.2>)', 'tuple: (<Integer: 1>, <Integer: 1>, <MachineReal: 0.4>)']
0.87:             ['MachineReals._machine_reals', 'tuple: (<MachineReal: 0.87>, <MachineReal: 0.94>, <Integer: 1>)']
0.94:             ['MachineReals._machine_reals', 'tuple: (<MachineReal: 0.87>, <MachineReal: 0.94>, <Integer: 1>)']
0.85:             ['MachineReals._machine_reals', 'tuple: (<Symbol: System`Brown>, <MachineReal: 0.85>)', 'tuple: (<Symbol: System`Pink>, <MachineReal: 0.85>)', 'tuple: (<Symbol: System`Red>, <MachineReal: 0.85>)']
0.9:              ['MachineReals._machine_reals', 'tuple: (<Symbol: System`Cyan>, <MachineReal: 0.9>)', 'tuple: (<Integer: 1>, <MachineReal: 0.9>, <MachineReal: 0.8>)']
0.88:             ['MachineReals._machine_reals', 'tuple: (<Symbol: System`Green>, <MachineReal: 0.88>)', 'tuple: (<Symbol: System`Purple>, <MachineReal: 0.88>)']
0.8:              ['MachineReals._machine_reals', 'tuple: (<Integer: 1>, <MachineReal: 0.9>, <MachineReal: 0.8>)']
1.0:              ['MachineReals._machine_reals', 'tuple: (<MachineReal: 1.0>, <MachineReal: 0.5>, <MachineReal: 0.5>)']
1.3:              ['MachineReals._machine_reals', 'tuple: (<MachineReal: 1.3>, <MachineReal: -2.4>, <Integer: 2>)']
-2.4:             ['MachineReals._machine_reals', 'tuple: (<MachineReal: 1.3>, <MachineReal: -2.4>, <Integer: 2>)']
2.0:              ['MachineReals._machine_reals']
743.0:            ['MachineReals._machine_reals', 'tuple: (<MachineReal: 743.0>, <Integer: -1>)', 'tuple: (<MachineReal: 743.0>, <Integer: -1>)']
0.0013458950201884253: ["(<class 'functools._lru_cache_wrapper'>, 140479778063360)", 'MachineReals._machine_reals']
0.16689098250336473: ["(<class 'functools._lru_cache_wrapper'>, 140479778063360)", "(<class 'list'>, 140479632004992)", 'MachineReals._machine_reals', "dict: {'System`x'}", "dict: {'System`y', 'System`x'}", "dict: {'x', 'y'}", "dict: {'evaluation', 'y', 'x'}", 'tuple: (<MachineReal: 0.16689098250336473>, <Expression: System`Times[1, System`Power[2, -1]]>)', 'tuple: (<MachineReal: 0.16689098250336473>, <Rational: 1/2>)', 'tuple: (<MachineReal: 0.16689098250336473>,)', 'tuple: (<MachineReal: 0.16689098250336473>, <Rational: 1/2>)', 'tuple: (<MachineReal: 0.16689098250336473>, <Rational: 1/2>)']
0.40852292775726157: ['MachineReals._machine_reals']
0.408523

---
None

list of entries and their references for the dict (<class 'dict'>, 140479948858880)
0.0:              ['MachineReals._machine_reals', "dict: {'expr', 'atom'}", "dict: {'replace', 'system', 'pattern'}", "dict: {'expr', 'atom'}", "dict: {'expr', 'atom'}", "dict: {'expr', 'atom'}", "dict: {'expr', 'atom'}", "dict: {'expr', 'atom'}", "dict: {'expr', 'atom'}", 'class / module', 'class / module', 'class / module: \nRepresentation of Numbers\n\nIntegers and Real numb', 'tuple: (<MachineReal: 0.0>,)', 'tuple: (<MachineReal: 0.0>,)', 'tuple: (<MachineReal: 0.0>,)', 'tuple: (<MachineReal: 0.0>,)', 'tuple: (<Symbol: System`z>, <MachineReal: 0.0>)', 'tuple: (<MachineReal: 0.0>,)', 'tuple: (<MachineReal: 0.0>,)']
-3.0:             ['MachineReals._machine_reals', "dict: {'original', 'hash', '_value'}", "dict: {'replace', 'system', 'pattern'}", "class / module: \n    <dl>\n      <dt>'$SystemTimeZone'\n      <dd> g", "class / module: \n    <dl>\n      <dt>'$TimeZone'\n      <dd> gives t"]
0.6:              ['MachineReals._machine_reals', 'class / module: \nEvaluation routines for 2D plotting.\n\nThese routi', 'tuple: (<MachineReal: 0.6>, <MachineReal: 0.4>, <MachineReal: 0.2>)']
0.2:              ['MachineReals._machine_reals', 'class / module: \nEvaluation routines for 2D plotting.\n\nThese routi', 'tuple: (<MachineReal: 0.6>, <MachineReal: 0.4>, <MachineReal: 0.2>)', 'tuple: (<Symbol: System`i>, <Integer: 2>, <MachineReal: 0.2>)', 'tuple: (<Symbol: System`i>, <Symbol: System`r>, <MachineReal: 0.2>)']
-10.0:            ['MachineReals._machine_reals']
-1.0:             ['MachineReals._machine_reals']
0.5:              ['MachineReals._machine_reals', 'tuple: (<MachineReal: 0.5>,)', 'tuple: (<MachineReal: 0.5>,)', 'tuple: (<Integer: 1>, <MachineReal: 0.5>, <Integer: 0>)', 'tuple: (<MachineReal: 1.0>, <MachineReal: 0.5>, <MachineReal: 0.5>)', 'tuple: (<MachineReal: 0.5>, <Integer: 0>, <MachineReal: 0.5>)', 'tuple: (<Symbol: System`n>, <MachineReal: 0.5>)', 'tuple: (<Symbol: System`n>, <MachineReal: 0.5>)', 'tuple: (<Symbol: System`n>, <MachineReal: 0.5>)', 'tuple: (<Symbol: System`n>, <MachineReal: 0.5>)']
0.4:              ['MachineReals._machine_reals', 'tuple: (<MachineReal: 0.6>, <MachineReal: 0.4>, <MachineReal: 0.2>)', 'tuple: (<Integer: 1>, <Integer: 1>, <MachineReal: 0.4>)']
0.87:             ['MachineReals._machine_reals', 'tuple: (<MachineReal: 0.87>, <MachineReal: 0.94>, <Integer: 1>)']
0.94:             ['MachineReals._machine_reals', 'tuple: (<MachineReal: 0.87>, <MachineReal: 0.94>, <Integer: 1>)']
0.85:             ['MachineReals._machine_reals', 'tuple: (<Symbol: System`Brown>, <MachineReal: 0.85>)', 'tuple: (<Symbol: System`Pink>, <MachineReal: 0.85>)', 'tuple: (<Symbol: System`Red>, <MachineReal: 0.85>)']
0.9:              ['MachineReals._machine_reals', 'tuple: (<Symbol: System`Cyan>, <MachineReal: 0.9>)', 'tuple: (<Integer: 1>, <MachineReal: 0.9>, <MachineReal: 0.8>)']
0.88:             ['MachineReals._machine_reals', 'tuple: (<Symbol: System`Green>, <MachineReal: 0.88>)', 'tuple: (<Symbol: System`Purple>, <MachineReal: 0.88>)']
0.8:              ['MachineReals._machine_reals', 'tuple: (<Integer: 1>, <MachineReal: 0.9>, <MachineReal: 0.8>)']
1.0:              ['MachineReals._machine_reals', 'tuple: (<MachineReal: 1.0>, <MachineReal: 0.5>, <MachineReal: 0.5>)']
1.3:              ['MachineReals._machine_reals', 'tuple: (<MachineReal: 1.3>, <MachineReal: -2.4>, <Integer: 2>)']
-2.4:             ['MachineReals._machine_reals', 'tuple: (<MachineReal: 1.3>, <MachineReal: -2.4>, <Integer: 2>)']
2.0:              ['MachineReals._machine_reals']
743.0:            ['MachineReals._machine_reals', 'tuple: (<MachineReal: 743.0>, <Integer: -1>)', 'tuple: (<MachineReal: 743.0>, <Integer: -1>)']
0.0013458950201884253: ["(<class 'functools._lru_cache_wrapper'>, 140479778063360)", 'MachineReals._machine_reals']
0.16689098250336473: ["(<class 'functools._lru_cache_wrapper'>, 140479778063360)", 'MachineReals._machine_reals', 'tuple: (<MachineReal: 0.16689098250336473>, <Rational: 1/2>)']
0.40852292775726157: ["(<class 'functools._lru_cache_wrapper'>, 140479778063360)", 'MachineReals._machine_reals', "dict: {'replace', 'system', 'pattern'}"]
1.

```

We see that three of the items (```2.```,``` -1.``` and ``` -10`.``) just have instances inside the class instances dict. ```743.``` are stored in the input expression, while the intermediate values (```0.16689098250336473```,```0.0013458950201884253``` ) were stored also in the lru_cache.

Reply:
It would be great to have a handler that allows to flush all the lru_cache. Then, we could use it in ```Share``` to remove all these instances when more memory is needed.

Reply:
I am not getting that much insight from the above.  We got into this because there was a fear that there would be all of these items in a dictionary cache and that would cause problems. 

That benchmark doesn't seem to say anything one way or the other about this concern. 

My claim is that all of these items are around in Python's object store any way and are not likely to be garbage collected. 

The lru cache is removed in this branch, because we are going with the home-grown caching approach since we  need `__new__()` to compute object hash values and in some cases are hoping that we can  canonicalize numeric values. 

We could put lru_cache back in and remove the homegrown cache, but unless this is causing new major problems (and I don't see that yet), this just isn't something I want to focus on right now, and I don't think we should be focusing on this.

The point of departure here had to do with seeing that the hash value is computed many times when it needs to only do it once on object creation.  Then there was the question of how often object creation for a given numeric value occurs. 

A more major concern I have is organizing the code base to be more like a conventional interpreter, and one that is better documented and understandable.

Reply:
> * At some point, what we need to do is to improve `Share` to clean both the lru cache and the unused entries in the class dictionary.
> * It would be also good to introduce classes that handle large arrays without creating MachineReal objects, at least for the intermediate steps.

In https://github.com/Mathics3/mathics-core/pull/664/commits/55ced2c7f408149ff8a425841067f1ba70644a31 I made an attempt to note the connection in this code versus `@lru_cache` and its relation with Builtin `Share[]`. If this isn't sufficient, feel free to add more. 

If you want to add notes or ideas in the ideas section, or open this as a feature for the future, please go ahead.

It is not my intent is to stifle ideas. It is just that right now, we are just a long ways away from handling  something like `Share[]` in any serious - we don't have a handle on normal evaluation.

In my opinion a big part of this is that the code isn't organized in a way where we can reasonably make sense of it and do the kinds of things that are typically done to speed up evaluation.  Let's start along the lines that interpreters typically use to speed things up before inventing or relying on custom tweaking to get there.

It has been noted, and I have noted for a while that yes, we need a way to handle large lists of numeric literals, and be able to operate on such large lists without massive conversions. It is not that I don't have ideas about this, it is again a matter of coming up with a plan of what to tackle next and focusing on that. 

If after the next release, if that's the thing you'd like to focus on instead of the makebox, boxing rules and formatting, I am okay for doing that.

Reply:
> In [55ced2c](https://github.com/Mathics3/mathics-core/commit/55ced2c7f408149ff8a425841067f1ba70644a31) I made an attempt to note the connection in this code versus `@lru_cache` and its relation with Builtin `Share[]`. If this isn't sufficient, feel free to add more.

By now, I think this is enough. My point with the comment was just to note that in the future we need to handle the overhead of storing all the used numbers in some way.


> 
> It is not my intent is to stifle ideas. It is just that right now, we are just a long ways away from handling something like `Share[]` in any serious - we don't have a handle on normal evaluation.
> 

Sure, I think we do not need to worry too much about this until we really are in the position of working with very large sets of numbers.


> In my opinion a big part of this is that the code isn't organized in a way where we can reasonably make sense of it and do the kinds of things that are typically done to speed up evaluation. Let's start along the lines that interpreters typically use to speed things up before inventing or relying on custom tweaking to get there.
> 

> It has been noted, and I have noted for a while that yes, we need a way to handle large lists of numeric literals, and be able to operate on such large lists without massive conversions. It is not that I don't have ideas about this, it is again a matter of coming up with a plan of what to tackle next and focusing on that.
> 
> If after the next release, if that's the thing you'd like to focus on instead of the makebox, boxing rules, and formatting, I am okay for doing that.

What I really want to have is a more clear API pattern, in a way that improving Boxes or the handling of large lists can be done outside Mathics-core. In this sense, I feel more comfortable continuing with the formatting refactor. In the process, we can discuss on a more concrete basis how the API and the organization of the code should be done.